client/src/components/AuditAdmin/Progress/Progress.test.tsx

```diff
@@ -738,6 +786,69 @@ describe('Progress screen', () => {
     })
   })
 
+  it('shows the detail modal with JA file download buttons after a batch audit starts', async () => {
+    const expectedCalls = [
+      aaApiCalls.getMapData,
+      jaApiCalls.getAuditBoards(auditBoardMocks.unfinished),
+      apiCalls.getBatches(batchesMocks.emptyInitial),
+      apiCalls.getBatches(batchesMocks.emptyInitial),
+      apiCalls.getJurisdictionContests({ contests: contestMocks.oneTargeted }),
+    ]
+    const blob = new Blob()
+    ;(pdf as jest.Mock).mockImplementation((doc: JSX.Element) => ({
+      toBlob: jest.fn(() => {
+        render(doc)
+        return Promise.resolve(blob)
+      }),
+    }))
+
+    await withMockFetch(expectedCalls, async () => {
+      render(
+        <QueryClientProvider client={queryClient}>
+          <Progress
+            auditSettings={auditSettings.batchComparisonAll}
+            jurisdictions={jurisdictionMocks.oneComplete}
+            round={roundMocks.singleComplete[0]}
+          />
+        </QueryClientProvider>
+      )
+
+      // Open detail modal
+      userEvent.click(
+        await screen.findByRole('button', { name: 'Jurisdiction 1' })
+      )
+      const modal = screen
+        .getByRole('heading', { name: 'Jurisdiction 1' })
+        .closest('div.bp3-dialog')! as HTMLElement
+      await within(modal).findByRole('heading', {
+        name: 'Round 1 Data Entry',
+      })
+
+      userEvent.click(
+        await within(modal).findByRole('button', {
+          name: /Download Aggregated Batch Retrieval List/,
+        })
+      )
+      expect(window.open).toHaveBeenCalledWith(
+        '/api/election/1/jurisdiction/jurisdiction-id-1/round/round-1/batches/retrieval-list'
+      )
+
+      await act(async () => {
```

Comment:
What's this `await act(...)` for? I thought `userEvent.click` had `act` built in

Reply:
Ooh very interesting. As I was building this test out and this click was its last line (because I hadn't yet written the assertions that came after), I was receiving the following warning from Jest:

<img width="1265" alt="warning" src="https://user-images.githubusercontent.com/12616928/167960688-ad4d0855-349f-4a45-963b-dfd8bd5594b5.png">

Adding the `act` resolved the warning. I assumed it was needed because of the newly added `setImmediate` triggering state updates with a delay.

But now that I have some assertions after this line, the `act` no longer appears to be necessary. Removing the assertions afterwards results in the above error resurfacing.

Will remove the `act` calls but still curious to find an explanation for the above behavior (a task for tomorrow)

Reply:
There's a short explanation of the act warning here: https://testing-library.com/docs/react-testing-library/faq/. And it also links to a longer blog post with a good explanation. It basically means that there was code trying to make React state updates after the test already finished.

Reply:
Ah perfect, thanks for that!